### ISSUE TITLE
Need to write a proper `system-agent` uninstall script

### ISSUE THREAD

**Issue by Oats87** [rancher/rancher#37283]
The system-agent today currently installs via a posix-compliant shell script found at https://github.com/rancher/system-agent/blob/main/install.sh

We need to create a corresponding uninstall script that removes the system-agent from the node. 

Notably, it should just "ctrl-z" the actions that the install script performs.

It should clean up the stock directories for the system-agent i.e. 

```
/etc/rancher/agent
/var/lib/rancher/agent
/etc/systemd/system/rancher-system-agent.service
/etc/systemd/system/rancher-system-agent.env
/usr/local/bin/rancher-system-agent
- OR - 
/opt/rancher-system-agent/bin/rancher-system-agent
```

It should accept the same environment variables as the install script to handle the case where the user changes the directories the script installs to.

Bonus points if we install the uninstall script along with the system-agent.

**Comment by phillipsj** [comment]
@jakefhyde @rosskirkpat @luthermonson @sirredbeard just pinging you all so we don't forget that we need this for Windows too.

**Comment by sirredbeard** [comment]
@jakefhyde If this is implemented for Linux, it needs to be implemented for Windows too.

**Comment by snasovich** [comment]
@sirredbeard @phillipsj , we're not too familiar with the way it work on Windows, so would like your assistance for this. Do you want a separate issue created elsewhere for this?

@Oats87 , would be great if you could also clarify the "ask" in this ticket as it's not clear what's the purpose or intended future use of the requested script.

**Comment by phillipsj** [comment]
Happy to help.

**Comment by snasovich** [comment]
@sirredbeard @phillipsj, added https://github.com/rancher/windows/issues/171 - please add labels / status as necessary.

Ideally, we want this script done ready for v2.6.5 release, but it's not part of "code" per se so it can be done up until the release date. @jakefhyde, FYI.

**Comment by phillipsj** [comment]
@snasovich both issues have the 2.6.6 milestone. Do we want this for 2.6.5 or 2.6.6?

**Comment by sirredbeard** [comment]
@snasovich Doesn't the script need to go into the RKE2 release, which has already been cut for April?

**Comment by snasovich** [comment]
@sirredbeard , this issue is to come up with a script to cleaning up `system-agent` manually.
This is in addition to removing RKE2 which is detailed here: https://docs.rke2.io/install/linux_uninstall/#tarball-method.
In a long run we probably want this done automatically by Rancher as part of deleting Custom RKE2 clusters, similar to how cleanup is done for Custom RKE1 clusters. So right now there is no need to "bake" this script into anything - just document (ideally we should include it with `system-agent` install similar to how RKE2 does it, but it's too late for that for 2.6.5 as you correctly pointed out).

**Comment by sirredbeard** [comment]
@phillipsj Is mostly done with the script and will be testing. It can be added to docs when Linux script is added to the docs.

cc @snasovich

**Comment by jakefhyde** [comment]
### Root cause

System agent  did not have an uninstall script
 
### What was fixed, or what changes have occurred

An uninstall script was installed alongside system agent when the install script is used.
 
### Areas or cases that should be tested

Airgap and non-airgap clusters should have no problem installing the uninstall script
The uninstall script should uninstall everything related to system agent
 
### What areas could experience regressions?

N/A
 
### Are the repro steps accurate/minimal?
:heavy_check_mark:

**Comment by jakefhyde** [comment]
This issue is currently blocked by https://github.com/rancher/rancher/issues/38438. Once the associated PR (https://github.com/rancher/system-agent/pull/88) is merged, this issue can be tested on airgap.

**Comment by jameson-mcghee** [comment]
Additional Information from @jakefhyde:
"the system-agent install script will install it automatically when it installs the system-agent binary. For a custom cluster, you'll remove the node from the cluster and have to manually ssh in and run the script yourself. It will remove all related system-agent resources:
**/var/lib/rancher/agent
/etc/rancher/agent
/usr/local/bin/rancher-system-agent
/opt/rancher-system-agent (if /usr/local is read-only)
/etc/systemd/system/rancher-system-agent.service
/etc/systemd/system/rancher-system-agent.env**
The install script and the uninstall script will be left behind in /usr/local/bin, but everything else should be removed."
cc @slickwarren @sowmyav27

**Comment by slickwarren** [comment]
so to summarize, here is our tentative test plan:
uninstall script is at /usr/local/bin The install script and the uninstall script will be left behind in /usr/local/bin, but everything else should be removed." with the name rancher-system-agent-uninstall.sh
test cases:
* verify that the uninstall script is available for the following rke2 cluster types:
  * node driver clusters 
  * custom clusters
* verify for custom clusters that the uninstall script works by doing the following (not necessary to test for node driver clusters because scaling down/removing a node from a node driver will remove the entire node):
  * delete the node from rancher
     * verify that the following resources still exist on the deleted node:
```
/var/lib/rancher/agent
/etc/rancher/agent
/usr/local/bin/rancher-system-agent
/opt/rancher-system-agent (if /usr/local is read-only)
/etc/systemd/system/rancher-system-agent.service
/etc/systemd/system/rancher-system-agent.env
```

* after node is removed on rancher side, ssh into the node and run the uninstall script. 
    * verify that the previously mentioned files are removed

edge cases:
* making sure it works on different role nodes (etcd, controlplane, worker).
* reusing a previously used node in a new cluster to verify that everything was cleaned up to the point that it is a useful node again. 
* There might be some other edge cases we can think of.

@jakefhyde could you clarify one point for us? In the original comment:
```
It should accept the same environment variables as the install script to handle the case where the user changes the directories the script installs to.
```
What vars to add to change the directory for install/uninstall scripts?

**Comment by jakefhyde** [comment]
@slickwarren https://github.com/rancher/system-agent/blob/6a89e30879b60966471bc52d4c499adfef6f65be/install.sh#L16-L21

So if you were to create an rke2 cluster with the env vars `CATTLE_AGENT_CONFIG_DIR=/tmp/rancher` for example, running the uninstall script with `CATTLE_AGENT_CONFIG_DIR=/tmp/rancher rancher-system-agent-uninstall.sh` should properly clean it up. While in theory everything should work if those vars are set, if they don't it would be a separate issue.

**Comment by slickwarren** [comment]
tested in airgap on v2.6.7-rc6 (awaiting other section of testing in non-airgap setup):
* verify that the uninstall script is available for the following rke2 cluster types:
  * custom clusters  -- pass (installed once node is registered with the cluster)
* verify for custom clusters that the uninstall script works by doing the following (not necessary to test for node driver clusters because scaling down/removing a node from a node driver will remove the entire node):
  * delete the node from rancher
    * verify that the following resources still exist on the deleted node: (see https://github.com/rancher/rancher/issues/37283#issuecomment-1204365810) -- pass
  * after node is removed on rancher side, ssh into the node and run the uninstall script.
    * verify that the previously mentioned files are removed -- pass
  * run rke2 uninstall script
  * register node in a new cluster -- pass
* create a  custom cluster with env. var `CATTLE_AGENT_CONFIG_DIR=/tmp/rancher` set and register a node
  * verify that the cattle agent config dir is properly updated based on the user vars -- pass
* follow the main steps from this thread, but specifying the env. var when running the agent-uninstall scripts
  * verify that the cattle agent uninstall is properly removing the following, in addition to the cattle_agent_config_dir based on the user vars -- pass
```
/var/lib/rancher/agent
/etc/rancher/agent
/usr/local/bin/rancher-system-agent
/opt/rancher-system-agent (if /usr/local is read-only)
/etc/systemd/system/rancher-system-agent.service
/etc/systemd/system/rancher-system-agent.env
```

**Comment by jameson-mcghee** [comment]
Tested in non-airgap on v2.6.7-rc6 using custom and EC2 downstream clusters. During testing all files that were removed as intended. An arbitrary error message was observed, but it did not effect the outcome of the testing. Therefore non-airgapped testing is done.
cc @slickwarren @jakefhyde 

**/usr/local/bin# bash rancher-system-agent-uninstall.sh**
Removed /etc/systemd/system/multi-user.target.wants/rancher-system-agent.service.
Failed to reset failed state of unit rancher-system-agent.service: Unit rancher-system-agent.service not loaded.

**/usr/local/bin# ls**
rke2  rke2-killall.sh  rke2-uninstall.sh

**/var/lib/rancher# ls**
rke2

**/etc/rancher# ls**
node  rke2

**/etc/systemd/system# ls**
 cloud-final.service.wants                default.target.wants     graphical.target.wants    multipath-tools.service          rescue.target.wants                       snap-core18-2538.mount   snap-snapd-16292.mount                           snap.lxd.daemon.unix.socket   syslog.service
 cloud-init.target.wants                  emergency.target.wants   iscsi.service             network-online.target.wants     'snap-amazon\x2dssm\x2dagent-3552.mount'   snap-core20-1587.mount   snap.amazon-ssm-agent.amazon-ssm-agent.service   sockets.target.wants          timers.target.wants
 dbus-org.freedesktop.resolve1.service    final.target.wants       mdmonitor.service.wants   open-vm-tools.service.requires  'snap-amazon\x2dssm\x2dagent-5656.mount'   snap-lxd-19647.mount     snap.lxd.activate.service                        sshd.service                  vmtoolsd.service
 dbus-org.freedesktop.timesync1.service   getty.target.wants       multi-user.target.wants   paths.target.wants               snap-core18-1988.mount                    snap-lxd-22753.mount     snap.lxd.daemon.service                          sysinit.target.wants
